extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeJavaScriptPlayerManager.java

```diff
@@ -0,0 +1,334 @@
+
+    @Nonnull
+    private static final Map<String, String> CACHED_THROTTLING_PARAMETERS = new HashMap<>();
+
+    private static String cachedJavaScriptPlayerCode;
+
+    @Nullable
+    private static Integer cachedSignatureTimestamp;
+    @Nullable
+    private static String cachedSignatureDeobfuscationFunction;
+    @Nullable
+    private static String cachedThrottlingDeobfuscationFunctionName;
+    @Nullable
+    private static String cachedThrottlingDeobfuscationFunction;
+
+    @Nullable
+    private static ParsingException throttlingDeobfFuncExtractionEx;
+    @Nullable
+    private static ParsingException sigDeobFuncExtractionEx;
+    @Nullable
+    private static ParsingException sigTimestampExtractionEx;
+
+    private YoutubeJavaScriptPlayerManager() {
+    }
+
+    /**
+     * Get the signature timestamp of the base JavaScript player file.
+     *
+     * <p>
+     * A valid signature timestamp sent in the payload of player InnerTube requests is required to
+     * get valid stream URLs on HTML5 clients for videos which have obfuscated signatures.
+     * </p>
+     *
+     * <p>
+     * The base JavaScript player file will fetched if it is not already done.
+     * </p>
+     *
+     * <p>
+     * The result of the extraction is cached until {@link #clearAllCaches()} is called, making
+     * subsequent calls faster.
+     * </p>
+     *
+     * @param videoId the video ID used to get the JavaScript base player file (an empty one can be
+     *                passed, even it is not recommend in order to spoof better official YouTube
+     *                clients)
+     * @return the signature timestamp of the base JavaScript player file
+     * @throws ParsingException if the extraction of the base JavaScript player file or the
+     * signature timestamp failed
+     */
+    @Nonnull
+    public static Integer getSignatureTimestamp(@Nonnull final String videoId)
+            throws ParsingException {
+        // Return the cached result if it is present
+        if (cachedSignatureTimestamp != null) {
+            return cachedSignatureTimestamp;
+        }
+
+        // If the signature timestamp has been not extracted on a previous call, this mean that we
+        // will fail to extract it on next calls too if the player code has been not changed
+        // Throw again the corresponding stored exception in this case to improve performance
+        if (sigTimestampExtractionEx != null) {
+            throw sigTimestampExtractionEx;
+        }
+
+        extractJavaScriptCodeIfNeeded(videoId);
+
+        try {
+            cachedSignatureTimestamp = Integer.valueOf(
+                    YoutubeSignatureUtils.getSignatureTimestamp(cachedJavaScriptPlayerCode));
+        } catch (final ParsingException e) {
+            // Store the exception for future calls of this method, in order to improve performance
+            sigTimestampExtractionEx = e;
+            throw e;
+        }
```

Comment:
What if `Integer.valueOf` fails? Can't you just catch any exception with `Throwable` in all places?

Reply:
> What if `Integer.valueOf` fails?

I need to handle at least `NumberFormatException`s here indeed, good catch!

> Can't you just catch any exception with `Throwable` in all places?

Is it really a good idea to catch all errors, including ones which could come from the JVM (i.e. instances of the `Error` class)? I don't think so, but I am not sure.